pkg/querier/querier.go

```diff
@@ -377,6 +379,120 @@ func (q *Querier) tailDisconnectedIngesters(ctx context.Context, req *logproto.T
 	return reconnectClientsMap, nil
 }
 
+// Series fetches any matching series for a list of matcher sets
+func (q *Querier) Series(ctx context.Context, req *logproto.SeriesRequest) (*logproto.SeriesResponse, error) {
+	userID, err := user.ExtractOrgID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = q.validateQueryTimeRange(userID, &req.Start, &req.End); err != nil {
+		return nil, err
+	}
+
+	// Enforce the query timeout while querying backends
+	ctx, cancel := context.WithDeadline(ctx, time.Now().Add(q.cfg.QueryTimeout))
+	defer cancel()
+
+	series, errs := make(chan [][]logproto.SeriesIdentifier, 2), make(chan error, 2)
+
+	go func() {
+		// fetch series identifiers from ingesters
+		resps, err := q.forAllIngesters(func(client logproto.QuerierClient) (interface{}, error) {
+			return client.Series(ctx, req)
+		})
+		if err != nil {
+			errs <- err
+			return
+		}
+		var acc [][]logproto.SeriesIdentifier
+		for _, resp := range resps {
+			acc = append(acc, resp.response.(*logproto.SeriesResponse).Series)
+		}
+		series <- acc
+	}()
+
+	go func() {
+		storeValues, err := q.seriesForMatchers(ctx, req.Start, req.End, req.GetGroups())
+		if err != nil {
+			errs <- err
+			return
+		}
+		series <- [][]logproto.SeriesIdentifier{storeValues}
+	}()
+
+	var sets [][]logproto.SeriesIdentifier
```

Comment:
There's a small problem here: If one of the goroutines fails, the other one will leak and be blocked forever waiting to write to the channel. 

I think the best way to fix this is just to store the last received error and return the error after the for loop, if it's non-nil.

I'm also finding this method a little too complex for me. Do we need to run the series query against the ingesters and store concurrently? Maybe the goroutines can be extracted out into methods to make this method a little shorter overall?

Reply:
They shouldn't block as they're both buffered channels with length of 2, which equals the # of times they'd be written to (1 for each goroutine). Additionally, doing it this way lets us return once we've seen the first error instead of waiting for all the downstream calls to finish in order to drain the channel and then returning the first error found. I'll move the channel code to it's own function -- let me know if this is insufficient or I'm missing something.

Reply:
Nope, sorry, missed that they were buffered.